### PR TITLE
[10.0][ADD] planned orders to a blanket order

### PR DESCRIPTION
**Pull request by Laurent-Corron**
allows to add planned orders to a blanket order

### CODE REVIEW

File: sale_blanket_order/models/blanket_orders.py [hunk 1]
```diff
@@ -307,7 +317,22 @@ def _compute_quantities(self):
...
+    @api.multi
+    @api.depends(
+        'planned_line_ids.next_order_date'
+    )
+    def _compute_next_planned_order(self):
+        for line in self:
+            planned_lines = line.planned_line_ids
+            if planned_lines:
+                next_planned_line = planned_lines.sorted(key=lambda r: r.next_order_date)[0]
+                for lines in next_planned_line:
```
Comment (sudhir-erpharbor) [inline sale_blanket_order/models/blanket_orders.py]: What is the use of the for loop here? I think it needs to be removed.

File: sale_blanket_order/models/planned_blanket_orders.py [hunk 2]
```diff
@@ -0,0 +1,73 @@
+# coding: utf-8
+# Copyright 2018 ACSONE SA/NV
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+from odoo import fields, models, api, _
```
Comment (sudhir-erpharbor) [inline sale_blanket_order/models/planned_blanket_orders.py]: Import statement should be in alphabetical order

File: sale_blanket_order/models/planned_blanket_orders.py [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+    next_order_date = fields.Date(string='next order date', required=True)
+    line_ids = fields.One2many(
+        'sale.blanket.planned.order.line', 'planned_order_id', string='Order lines',
+        copy=True)
+
+    @api.multi
+    def create_order(self):
+        for rec in self:
+            if fields.Date.today() == rec.next_order_date:
+                line_obj = self.env['sale.order.line']
```
Comment (sudhir-erpharbor) [inline sale_blanket_order/models/planned_blanket_orders.py]: Define object before the loop

File: sale_blanket_order/models/planned_blanket_orders.py [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+        'sale.blanket.order', required=True, ondelete='cascade')
+    next_order_date = fields.Date(string='next order date', required=True)
+    line_ids = fields.One2many(
+        'sale.blanket.planned.order.line', 'planned_order_id', string='Order lines',
+        copy=True)
+
+    @api.multi
+    def create_order(self):
+        for rec in self:
+            if fields.Date.today() == rec.next_order_date:
```
Comment (sudhir-erpharbor) [inline sale_blanket_order/models/planned_blanket_orders.py]: Better to use filtered.

    for rec self.filtered(lambda r: r.next_order_date == fields.Date.today()):

File: sale_blanket_order/models/planned_blanket_orders.py [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+                    'currency_id': rec.blanket_order_id.currency_id.id,
+                    'order_line': order_lines,
+                    'pricelist_id': (rec.blanket_order_id.pricelist_id.id
+                                     if rec.blanket_order_id.pricelist_id
+                                     else False),
+                    'payment_term_id': (rec.blanket_order_id.payment_term_id.id
+                                        if rec.blanket_order_id.payment_term_id
+                                        else False),
+                })
+                self.env['sale.order'].create(order_vals)
```
Comment (sudhir-erpharbor) [inline sale_blanket_order/models/planned_blanket_orders.py]: Define obj before loop and not inside the loop

File: sale_blanket_order/models/blanket_orders.py [hunk 6]
```diff
@@ -307,7 +317,22 @@ def _compute_quantities(self):
             line.ordered_qty = sum(l.product_uom_qty for l in sale_lines)
             line.invoiced_qty = sum(l.qty_invoiced for l in sale_lines)
             line.delivered_qty = sum(l.qty_delivered for l in sale_lines)
-            line.remaining_qty = line.original_qty - line.ordered_qty
+            line.planned_qty = sum(l.next_order_qty for l in line.planned_line_ids)
+            line.remaining_qty = line.original_qty - line.ordered_qty - line.planned_qty
+
+    @api.multi
```
Comment (sudhir-erpharbor) [inline sale_blanket_order/models/blanket_orders.py]: No need of api.multi as api.depends will do the job

**Review by sudhir-erpharbor — CHANGES_REQUESTED** [review]

**Comment by sudhir-erpharbor** [comment]
Please fix travis errors.

**Comment by sbidoul** [comment]
@Laurent-Corron is covered by the Acsone ECLA. The CLA bot should notice it soon now.

**Comment by sbidoul** [comment]
It would be nice to update the README fragments with a word about the new feature and how to use it. That will help both reviewers and future users understanding it.

File: sale_blanket_order/models/blanket_orders.py (hunk 1 above)
Comment (Laurent-Corron) [reply sale_blanket_order/models/blanket_orders.py]: Sorry was used for debugging, my bad!
It has been removed

File: sale_blanket_order/models/planned_blanket_orders.py (hunk 4 above)
Comment (Laurent-Corron) [reply sale_blanket_order/models/planned_blanket_orders.py]: fixed!

File: sale_blanket_order/models/planned_blanket_orders.py (hunk 5 above)
Comment (Laurent-Corron) [reply sale_blanket_order/models/planned_blanket_orders.py]: fixed!

**Comment by Laurent-Corron** [comment]
@Cedric-Pigeon

File: sale_blanket_order/models/planned_blanket_orders.py (hunk 3 above)
Comment (Laurent-Corron) [reply sale_blanket_order/models/planned_blanket_orders.py]: It was moved.

File: sale_blanket_order/models/blanket_orders.py [hunk 7]
```diff
@@ -36,6 +36,9 @@ def _default_company(self):
     lines_ids = fields.One2many(
         'sale.blanket.order.line', 'order_id', string='Order lines',
         copy=True)
+    planned_order_ids = fields.One2many(
+        'sale.blanket.planned.order', 'blanket_order_id', string='planned Orders',
+        copy=True)
```
Comment (Cedric-Pigeon) [inline sale_blanket_order/models/blanket_orders.py]: Would put copy=False. If we duplicate an expired blanket order, no need to copy the planification

File: sale_blanket_order/models/blanket_orders.py [hunk 8]
```diff
@@ -181,16 +184,22 @@ class BlanketOrderLine(models.Model):
...
     order_id = fields.Many2one(
         'sale.blanket.order', required=True, ondelete='cascade')
+    planned_line_ids = fields.One2many(
+        'sale.blanket.planned.order.line', 'blanket_order_line_id', string='Planned order lines')
     product_id = fields.Many2one(
         'product.product', string='Product', required=True)
     product_uom = fields.Many2one(
         'product.uom', string='Unit of Measure', required=True)
     price_unit = fields.Float(string='Price', required=True)
+    next_order_qty = fields.Float(string='next planned order Qty', compute='_compute_next_planned_order', store=True)
```
Comment (Cedric-Pigeon) [inline sale_blanket_order/models/blanket_orders.py]: please put the first letter in capital for all words in string attribute.

File: sale_blanket_order/models/blanket_orders.py [hunk 9]
```diff
@@ -181,16 +184,22 @@ class BlanketOrderLine(models.Model):
...
+    next_order_date = fields.Date(string='next order date', compute='_compute_next_planned_order', store=True)
```
Comment (Cedric-Pigeon) [inline sale_blanket_order/models/blanket_orders.py]: Same remark

File: sale_blanket_order/models/blanket_orders.py [hunk 10]
```diff
@@ -181,16 +184,22 @@ class BlanketOrderLine(models.Model):
...
     original_qty = fields.Float(
         string='Original quantity', required=True, default=1,
         digits=dp.get_precision('Product Unit of Measure'))
     ordered_qty = fields.Float(
         string='Ordered quantity', compute='_compute_quantities', store=True)
+    planned_qty = fields.Float(
+        string='Planned quantity', compute='_compute_quantities', store=True)
```
Comment (Cedric-Pigeon) [inline sale_blanket_order/models/blanket_orders.py]: Same remark

File: sale_blanket_order/models/planned_blanket_orders.py [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+
+
+class BlanketPlannedOrder(models.Model):
+    _name = 'sale.blanket.planned.order'
+    _description = 'Blanket Planned Order'
+    blanket_order_id = fields.Many2one(
+        'sale.blanket.order', required=True, ondelete='cascade')
+    next_order_date = fields.Date(string='next order date', required=True)
+    line_ids = fields.One2many(
+        'sale.blanket.planned.order.line', 'planned_order_id', string='Order lines',
```
Comment (Cedric-Pigeon) [inline sale_blanket_order/models/planned_blanket_orders.py]: Same remark

File: sale_blanket_order/models/planned_blanket_orders.py [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+            rec.unlink()
+
+
+class BlanketPlannedOrderLine(models.Model):
+    _name = 'sale.blanket.planned.order.line'
+    _description = 'Blanket Planned Order Line'
+
+    planned_order_id = fields.Many2one(
+        'sale.blanket.planned.order', ondelete='cascade')
+    next_order_qty = fields.Float(string='next order qty', required=True)
```
Comment (Cedric-Pigeon) [inline sale_blanket_order/models/planned_blanket_orders.py]: string is not needed

File: sale_blanket_order/models/planned_blanket_orders.py [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+    blanket_order_line_id = fields.Many2one(
+        'sale.blanket.order.line', readonly=True)
+    remaining_qty = fields.Float(string='Remaining quantity', related='blanket_order_line_id.remaining_qty')
```
Comment (Cedric-Pigeon) [inline sale_blanket_order/models/planned_blanket_orders.py]: string is not needed

File: sale_blanket_order/models/planned_blanket_orders.py [hunk 14]
```diff
@@ -0,0 +1,72 @@
...
+    product_id = fields.Many2one(string='product', related='blanket_order_line_id.product_id')
+    next_order_date = fields.Date(string='next order date', related='planned_order_id.next_order_date')
```
Comment (Cedric-Pigeon) [inline sale_blanket_order/models/planned_blanket_orders.py]: string is not needed

File: sale_blanket_order/wizard/create_sale_orders.py [hunk 15]
```diff
@@ -33,16 +33,17 @@ def _default_lines(self):
         }) for l in blanket_order.lines_ids]
         return lines
 
+    mode = fields.Char()
     blanket_order_id = fields.Many2one(
         'sale.blanket.order', default=_default_order, readonly=True)
     lines_ids = fields.One2many(
         'sale.blanket.order.wizard.line', 'wizard_id',
         string='Lines', default=_default_lines)
+    planned_date = fields.Date(string='planned date')
```
Comment (Cedric-Pigeon) [inline sale_blanket_order/wizard/create_sale_orders.py]: string not needed

File: sale_blanket_order/wizard/create_sale_orders.py [hunk 16]
```diff
@@ -95,6 +96,48 @@ def create_sale_order(self):
...
+    @api.multi
+    def create_planned_sale_order(self):
+        self.ensure_one()
+        order_lines = []
+        if self.planned_date > self.blanket_order_id.validity_date:
+            raise UserError(_('planned date must be before the validity date of the blanker order: %s') %
+                            self.blanket_order_id.validity_date)
+
+        for line in self.lines_ids:
+            if line.qty == 0.0:
```
Comment (Cedric-Pigeon) [inline sale_blanket_order/wizard/create_sale_orders.py]: please use float_compare method

File: sale_blanket_order/wizard/create_sale_orders.xml [hunk 17]
```diff
@@ -5,6 +5,10 @@
         <field name="model">sale.blanket.order.wizard</field>
         <field name="arch" type="xml">
             <form string="Create Sale Order">
+                <field name="mode" invisible="1"/>
+                <label for="planned_date" string="Planned date"/>
```
Comment (Cedric-Pigeon) [inline sale_blanket_order/wizard/create_sale_orders.xml]: "Date"

**Review by Cedric-Pigeon — CHANGES_REQUESTED** [review]

File: sale_blanket_order/wizard/create_sale_orders.py (hunk 16 above)
Comment (Laurent-Corron) [reply sale_blanket_order/wizard/create_sale_orders.py]: Done! Also changed it for the create_sale_order method in the wizard.

File: sale_blanket_order/models/blanket_orders.py (hunk 7 above)
Comment (Laurent-Corron) [reply sale_blanket_order/models/blanket_orders.py]: Done!

File: sale_blanket_order/models/blanket_orders.py (hunk 8 above)
Comment (Laurent-Corron) [reply sale_blanket_order/models/blanket_orders.py]: Done!

**Review by Cedric-Pigeon — APPROVED** [review]

**Review by rousseldenis — APPROVED** [review]

**Pull request closed without merging**